### ISSUE TITLE
Numerical Overflow while normalizing values with softmax

### ISSUE THREAD

**Issue by davidemandrini** [jpmml/jpmml-evaluator#66]
Hello,

while working with some logistic regression models I noticed that the current implementation for softmax normalisation can lead to overflow.

Softmax normalisation for a value `xi` in a vector `X = {x1, x2, ..., xn}`  is defined as:
`softmax(xi) = exp(xi) / ( exp(x1) + exp(x2) + ... + exp(xn) )`

In case the denominator is greater than `Double.MAX_VALUE`, a Java Double variable containing such sum will have value Infinite.
If the numerator is lower than `Double.MAX_VALUE`, the softmax value will be the results of a finite value divide by Infinite value, so 0.
If the numerator is greater than `Double.MAX_VALUE`, the softmax value will be the results of a Infinite value divide by Infinite value, so NaN. This behaviour can lead to wrong predictions.

As detailed in [Deep Learning textbook](http://www.deeplearningbook.org) by Ian Goodfellow, Yoshua Bengio and Aaron Courville, chapter 4, a solution to overflow problem is to calculate the softmax value as the following, with `x_M = max(x1, x2, ..., xn)`:
`softmax(xi) = exp(xi-x_M) / ( exp(x1-x_M) + exp(x2-x_M) + ... + exp(xn-x_M) )`

This solution do not reduce precision (the formulation with and without `x_M` are mathematically equivalent) and is able to avoid overflow.

Best regards,
Davide

----------------------------------
Book reference:
@book{Goodfellow-et-al-2016,
    title={Deep Learning},
    author={Ian Goodfellow and Yoshua Bengio and Aaron Courville},
    publisher={MIT Press},
    note={\url{http://www.deeplearningbook.org}},
    year={2016}
}

**Comment by vruusmann** [comment]
Is it a theoretical concern, or were you bitten by it in practice (eg. troubleshooting an unexpected `0`, NaN or Infinity prediction)?

Protection against numeric under/overflow seems necessary during model training. But is it really necessary during model deployment? The "best configuration" (which is exported into PMML document) should be numerically stable in most cases?

**Comment by davidemandrini** [comment]
I had this problem in a concrete usage, troubleshooting a unexpected NaN prediction.

I think we need protection against numeric under/overflow both at model training time and model deployment time.
The numerical values that go into the PMML document are not impacted by such change.

**Comment by vruusmann** [comment]
> I had this problem in a concrete usage, troubleshooting a unexpected NaN prediction.

It is assumed that the deployment dataset is within the "bounds" of the training dataset. If you didn't get any NaNs during model training, then it whould be impossible to get any NaNs during model deployment (provided that the above condition is met).

Does your PMML document specify field "bounds"? For example, do all the `DataField` elements that correspond to continuous fields have `Interval` child elements:
```xml
<DataField name="Sepal.Length" dataType="double" optype="continuous">
  <Interval leftMargin="4.0" rightMargin="9.0">
</DataField>
```

Otherwise, it might be the case that you're trying to apply the model to an invalid/out-of-domain data record (eg. `Sepal.Length = -1` or `Sepal.Length = 100`), that should be rejected.

**Comment by davidemandrini** [comment]
> Does your PMML document specify field "bounds"?

No, there are no bounds in my fields' declaration. All fields are like
`<DataField name="x1" optype="continuous" dataType="double" />`

> If you didn't get any NaNs during model training, then it would be impossible to get any NaNs during model deployment

I do not agree with this: model training can be performed with any library/code, then the coefficients can be stored in a PMML document (not necessary generated by JPMML code), evaluated with JPMML-Evaluator.
Model generation overflows are not always related to model evaluation overflows.

> Otherwise, it might be the case that you're trying to apply the model to an invalid/out-of-domain data record

My model is related to text analysis, with each feature representing a n-gram. Each feature value represents the count for the specific n-gram found for an input text.
In my failing test, the max count for a n-gram is 49, and this value seems reasonable.

I want to highlight once again that the optimisation I am proposing has no impact on evaluation results, it just avoids overflow.
In case of input values not causing overflow, the results with and without such optimisation are strictly the same (the new formulation is mathematically equivalent).

**Comment by davidemandrini** [comment]
I forget to mention that I created a pull request https://github.com/jpmml/jpmml-evaluator/pull/67 that solved my issue and can be used to test different scenario.

**Comment by vruusmann** [comment]
> In my failing test, the max count for a n-gram is 49, and this value seems reasonable.

Interesting point, that the failing test relates to "bag of words"-type classification models. There is some summation going on while evaluating the `TextIndex` element, which may be propagating "problematic" n-gram values to the "softmax" normalizer.

I know from my integration testing experience with Scikit-Learn and Apache Spark ML "bag-of-words"-type classification models that their predictions are slightly more difficult to reproduce than for other model types. So, there may be something else that needs improving.

> I am proposing has no impact on evaluation results, it just avoids overflow.

Have to take into consideration a noticeable performance penalty at the application level, because this safety check adds computation that is not needed for 99.99(99)% of data records.

No other model evaluator (I can speak about multi-class classification models in R, Scikit-Learn and Apache Spark ML) does this extra computation. So, there is also a possibility that JPMML-Evaluator predictions start to differ from the original framework predictions (as being "too precise").

> I forgot to mention that I created a pull request #67 that solved my issue and can be used to test different scenario.

I'm researching a similar accuracy issue with TensorFlow (TF) multi-class classification models. TF does summation using 32-bit floats, so the numeric error "develops" much faster and to a greater extent. Your patch does not fix my issue, because even after doing `exp(xi-x_M)`, the  minimum and maximum values differ more than seven or eight orders of magnitude, and the order of summation also needs to be taken into account.

In other words, thanks for pointing out this problem area, but I'm taking extra time to figure out an even better solution.

**Comment by vruusmann** [comment]
> My model is related to text analysis, with each feature representing a n-gram. Each feature value represents the count for the specific n-gram found for an input text.
> In my failing test, the max count for a n-gram is 49, and this value seems reasonable.

Can you check what does the original Scikit-Learn or Apache Spark ML model predict for this problematic data record? Specifically, does it also give you a NaN prediction, or not?

Suppose that the original model makes a NaN prediction. In that case, wouldn't you expect the JPMML-Evaluator library to also make a NaN prediction (instead of some "sensible" prediction)?

**Comment by davidemandrini** [comment]
> There is some summation going on while evaluating the TextIndex element, which may be propagating "problematic" n-gram values to the "softmax" normalizer.

I do not have any TextIndex element in my model. Each feature name is the n-gram itself:
```
<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<PMML xmlns="http://www.dmg.org/PMML-4_3" version="4.3">
	<Header>
		<Application/>
		<Timestamp>2017-06-15T15:45:54Z</Timestamp>
	</Header>
	<DataDictionary>
		<DataField name="targetField" optype="categorical" dataType="double">
			<Value value="aa"/>
			<Value value="bb"/>
			<Value value="cc"/>
		</DataField>
		<DataField name="ngram1" optype="continuous" dataType="double"/>
		<DataField name="ngram2" optype="continuous" dataType="double"/>
	</DataDictionary>
	<RegressionModel functionName="classification" normalizationMethod="softmax">
		<MiningSchema>
			<MiningField name="targetField" usageType="target"/>
			<MiningField name="ngram1"/>
			<MiningField name="ngram2"/>
		</MiningSchema>
		<RegressionTable intercept="0.0" targetCategory="aa">
			<NumericPredictor name="ngram1" coefficient="-0.5176923610318248"/>
			<NumericPredictor name="ngram2" coefficient="-0.035459800497820895"/>
		</RegressionTable>
		<RegressionTable intercept="0.0" targetCategory="bb">
			<NumericPredictor name="ngram1" coefficient="-0.5176923610318248"/>
			<NumericPredictor name="ngram2" coefficient="-0.035459800497820895"/>
		</RegressionTable>
		<RegressionTable intercept="0.0" targetCategory="cc">
			<NumericPredictor name="ngram1" coefficient="-0.5176923610318248"/>
			<NumericPredictor name="ngram2" coefficient="-0.035459800497820895"/>
		</RegressionTable>
	</RegressionModel>
</PMML>
```

> Have to take into consideration a noticeable performance penalty at the application level, because this safety check adds computation that is not needed for 99.99(99)% of data records.

The only "expensive" part of my fix is about computing the max between the values for each possible target class. I suppose that computing the max is a `O(n)` operation in Java, where `n` is the number of target categories (in the example above, we have `n=3` for aa/bb/cc. In my specific case I have `n=22`). This seems to be quite irrelevant, but I do not know the whole JPMML-Evaluator code, so I let you evaluate the performances of such change for the generic use case.

> No other model evaluator (I can speak about multi-class classification models in R, Scikit-Learn and Apache Spark ML) does this extra computation

Apache Spark ML does this operation in [LogisticRegression.scala](https://github.com/apache/spark/blob/v2.1.0/mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala) file, function `raw2probabilityInPlace `, line 855.

> Can you check what does the original Scikit-Learn or Apache Spark ML model predict for this problematic data record? Specifically, does it also give you a NaN prediction, or not?

I debugged both SparkML and JPMML-Evalutator prediction process: both calculate the exact same array of raw prediction values. Then SparkML normalises all of them correctly, thanks to the max scaling, whilst JPMML-Evalutator gives NaN for some values.

> Suppose that the original model makes a NaN prediction. In that case, wouldn't you expect the JPMML-Evaluator library to also make a NaN prediction (instead of some "sensible" prediction)?

Even if the original model gives NaN prediction, I would expect JPMML-Evaluator library to be precise. In my opinion JPMML-Evaluator should be indipendent from any model generator tool. Imagine the same exact model generated by two tools, one giving NaN and one not giving NaN. In this case what JPMML-Evaluator should do if the idea was to follow the tool behaviour?
